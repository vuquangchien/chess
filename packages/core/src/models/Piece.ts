import {PieceColor} from './PieceColor';
import {PieceType} from './PieceType';
export type FenName =
  | 'p'
  | 'r'
  | 'n'
  | 'b'
  | 'q'
  | 'k'
  | 'P'
  | 'R'
  | 'N'
  | 'B'
  | 'Q'
  | 'K';
export const PieceTypes: {[key: string]: PieceType} = {
  PAWN: 'pawn',
  ROOK: 'rook',
  KNIGHT: 'knight',
  BISHOP: 'bishop',
  QUEEN: 'queen',
  KING: 'king',
};
export const PieceColors: {[key: string]: PieceColor} = {
  WHITE: 'white',
  BLACK: 'black',
};
export class Piece {
  public color: PieceColor;
  public type: PieceType;
  private isMoved = false;

  constructor(type: PieceType, color: PieceColor) {
    this.color = color;
    this.type = type;
  }

  static fromName(name: FenName): Piece {
    const color =
      name === name.toUpperCase() ? PieceColors.WHITE : PieceColors.BLACK;
    const type = name.toLowerCase();
    const pieceType = {
      p: PieceTypes.PAWN,
      r: PieceTypes.ROOK,
      n: PieceTypes.KNIGHT,
      b: PieceTypes.BISHOP,
      q: PieceTypes.QUEEN,
      k: PieceTypes.KING,
    }[type];
    if (!pieceType) throw new Error(`Unknown piece type: ${type}`);
    return new Piece(pieceType, color);
  }

  public getColor(): PieceColor {
    return this.color;
  }

  public isPieceMoved(): boolean {
    return this.isMoved;
  }
  public setPieceMoved(): void {
    this.isMoved = true;
  }
  public getType(): PieceType {
    return this.type;
  }
  public getIcon(): string {
    const icons = {
      rook: {
        white: '♖',
        black: '♜',
      },
      knight: {
        white: '♘',
        black: '♞',
      },
      bishop: {
        white: '♗',
        black: '♝',
      },
      queen: {
        white: '♕',
        black: '♛',
      },
      king: {
        white: '♔',
        black: '♚',
      },
      pawn: {
        white: '♙',
        black: '♟',
      },
    };
    return icons[this.type][this.color];
  }
}
