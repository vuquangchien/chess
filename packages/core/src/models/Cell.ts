import {FenName, Piece} from './Piece';

function getCharFromNumber(num: number): string {
  return String.fromCharCode('a'.charCodeAt(0) + num);
}
export class Cell {
  public file: number;
  public rank: number;
  public piece: Piece | null;
  public color: string;
  public name: string;
  private highlighted = false;

  constructor(file: number, rank: number, piece: Piece | null) {
    this.file = file;
    this.rank = rank;
    this.piece = piece;
    this.color = (this.file + this.rank) % 2 === 1 ? 'white' : 'black';
    this.name = getCharFromNumber(this.file) + (this.rank + 1);
  }
  static fromName(name: string, pieceName?: FenName): Cell {
    const file = name.charCodeAt(0) - 97;
    const rank = parseInt(name.charAt(1), 10) - 1;
    return new Cell(file, rank, pieceName ? Piece.fromName(pieceName) : null);
  }
  public getPiece(): Piece | null {
    return this.piece;
  }

  setPiece(piece: Piece | null) {
    this.piece = piece;
  }

  highlight() {
    this.highlighted = true;
  }

  isHighlighted() {
    return this.highlighted;
  }

  unHighlight() {
    this.highlighted = false;
  }
}
