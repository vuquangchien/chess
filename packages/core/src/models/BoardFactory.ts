import {Board} from './Board';
import {FenRecord} from './fen/FenRecord';

export class BoardFactory {
  public static createEmptyBoard(): Board {
    return new Board();
  }
  public static createFromFen(fen: string): Board {
    const board = new Board();
    const fenRecord = new FenRecord(fen);
    const cells = fenRecord.piecePlacement.getCells();
    cells.forEach(cell => {
      board.getByCellName(cell.name).setPiece(cell.piece);
    });
    board.setCurrentPlayer(fenRecord.activeColor);
    return board;
  }
  public static createBoardWithPieces(): Board {
    return BoardFactory.createFromFen(
      'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w - - 0 1'
    );
  }
}
