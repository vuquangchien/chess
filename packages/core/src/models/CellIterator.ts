import {Board} from './Board';
import {Cell} from './Cell';
import {Rank} from './Rank';

export class RankIterator {
  private board: Board;
  private index: number;

  constructor(board: Board) {
    this.board = board;
    this.index = -1;
  }

  hasNext(): boolean {
    return this.index < 7;
  }

  currentIndex(): number {
    return this.index;
  }

  next(): Rank {
    this.index++;
    return this.current();
  }

  private current(): Rank {
    return this.board.getRanks()[this.index];
  }
}
export class CellIterator {
  private cells: Cell[];
  private index: number;

  constructor(cells: Cell[]) {
    this.cells = cells;
    this.index = -1;
  }

  hasNext(): boolean {
    return this.index < this.cells.length - 1;
  }

  next(): Cell {
    this.index++;
    return this.current();
  }

  private current(): Cell {
    return this.cells[this.index];
  }
}
