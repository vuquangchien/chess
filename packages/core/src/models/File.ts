import {Cell} from './Cell';
import {CellIterator} from './CellIterator';

export class File {
  cells: Cell[];

  constructor(cells: Cell[]) {
    this.cells = cells;
  }

  public getCellIterator(): CellIterator {
    return new CellIterator(this.cells);
  }
}
