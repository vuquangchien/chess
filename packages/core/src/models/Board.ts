import {Cell} from './Cell';
import {CellIterator, RankIterator} from './CellIterator';

import {File} from './File';
import {Rank} from './Rank';
import * as _ from 'lodash';
import {Move, MoveFactory} from './Move';
import {SamePlayerNotMoveTwicePolicy} from './policies/SamePlayerNotMoveTwicePolicy';
import {PieceNotEmptyPolicy} from './policies/PieceNotEmptyPolicy';
import {PawnMovePolicy} from './policies/PawnMovePolicy';
import {RookCapturePolicy, RookMovePolicy} from './policies/RookMovePolicy';
import {NotBlockedByFriendPiecePolicy} from './policies/NotBlockedByFriendPiecePolicy';
import {
  BishopMovePolicy,
  WhiteMoveFirstPolicy,
} from './policies/BishopMovePolicy';
import {KingMovePolicy} from './policies/KingMovePolicy';
import {KnightMovePolicy, QueenMovePolicy} from './policies/QueenMovePolicy';
import {PawnCapturePolicy} from './policies/PawnCapturePolicy';

export class Board {
  public cells: Cell[][];
  public moves: Move[] = [];
  constructor() {
    const files: Cell[][] = [];
    for (let rowIndex = 7; rowIndex >= 0; rowIndex--) {
      const file: Cell[] = [];
      for (let fileIndex = 0; fileIndex < 8; fileIndex++) {
        const cell = new Cell(fileIndex, rowIndex, null);
        file.push(cell);
      }
      files.push(file);
    }
    this.cells = files;
  }

  public getRankIterator(): RankIterator {
    return new RankIterator(this);
  }
  public getCell(file: number, rank: number): Cell {
    return this.cells[rank][file];
  }
  public getByCellName(name: string): Cell {
    const file = name.charCodeAt(0) - 97;
    const rank = 8 - parseInt(name.charAt(1), 10);
    return this.getCell(file, rank);
  }
  public getRanks(): Rank[] {
    return this.cells.map(cells => new Rank(cells));
  }

  public getFiles(): File[] {
    const ranks = [];
    for (let rankIndex = 0; rankIndex < 8; rankIndex++) {
      const cells = _.filter(
        _.flatten(this.cells),
        (cell: Cell) => cell.rank === rankIndex
      );
      ranks.push(new File(cells));
    }
    return ranks;
  }

  public validateMove(move: Move): void {
    const movePolicies = [
      new PieceNotEmptyPolicy(),
      new WhiteMoveFirstPolicy(this),
      new SamePlayerNotMoveTwicePolicy(this),
      new NotBlockedByFriendPiecePolicy(this),
      new PawnMovePolicy(),
      new RookMovePolicy(),
      new BishopMovePolicy(),
      new KingMovePolicy(),
      new QueenMovePolicy(),
      new KnightMovePolicy(),
    ];
    const capturePolicies = [
      new PawnCapturePolicy(move),
      new RookCapturePolicy(move),
    ];
    if (move.isSimpleMove())
      movePolicies.forEach(policy => {
        const error = policy.isMoveValid(move);
        if (error) {
          throw new Error(error);
        }
      });
    if (move.isCaptureMove())
      capturePolicies.forEach(policy => {
        const error = policy.isMoveValid(move);
        if (error) {
          throw new Error(error);
        }
      });
  }

  public movePiece(from: Cell, to: Cell): void {
    const move = MoveFactory.createMove(from, to);
    this.validateMove(move);

    this.moves.push(move);
    to.setPiece(from.piece);
    from.piece!.setPieceMoved();
    from.setPiece(null);
  }
  public movePieceByCellName(from: string, to: string): void {
    const fromCell = this.getByCellName(from);
    const toCell = this.getByCellName(to);
    this.movePiece(fromCell, toCell);
  }
  public getCellIterator(): CellIterator {
    return new CellIterator(_.flatten(this.cells));
  }
  public highlightPossibleMoves(cell: Cell) {
    const possibleMoves = this.getPossibleMoves(cell);
    possibleMoves.forEach(possibleMove => {
      possibleMove.highlight();
    });
  }
  public getPossibleMoves(cell: Cell): Cell[] {
    const possibleMoves: Cell[] = [];
    const it = this.getCellIterator();
    while (it.hasNext()) {
      const toCell = it.next();
      try {
        const move = MoveFactory.createMove(cell, toCell);
        this.validateMove(move);
        possibleMoves.push(toCell);
      } catch (e) {
        console.log(e);
      }
    }
    return possibleMoves;
  }

  highlightCell(cell: Cell) {
    cell.highlight();
  }

  unHighlightCell(cell: Cell) {
    cell.unHighlight();
  }

  resetHighlight() {
    const it = this.getCellIterator();

    while (it.hasNext()) {
      const cell = it.next();
      cell.unHighlight();
    }
  }
}
