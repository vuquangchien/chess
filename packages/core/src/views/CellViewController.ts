import {IViewController} from '../framework/IViewController';
import {Cell} from '../models/Cell';

export class CellViewController implements IViewController {
  public cell: Cell;

  constructor(cell: Cell) {
    this.cell = cell;
  }
}
