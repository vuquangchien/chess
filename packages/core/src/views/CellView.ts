import {CellViewController} from './CellViewController';
import {IEngine} from './IEngine';
import {EngineFactory} from './EngineFactory';
import {AbstractView} from './AbstractView';

const Icons = {
  rook: {
    white: '♖',
    black: '♜',
  },
  knight: {
    white: '♘',
    black: '♞',
  },
  bishop: {
    white: '♗',
    black: '♝',
  },
  queen: {
    white: '♕',
    black: '♛',
  },
  king: {
    white: '♔',
    black: '♚',
  },
  pawn: {
    white: '♙',
    black: '♟',
  },
};
export class CellView extends AbstractView<CellViewController> {
  controller: CellViewController;
  engine: IEngine;

  constructor(controller: CellViewController) {
    super(controller);
    this.controller = controller;
    this.engine = EngineFactory.createEngine('ascii');
  }

  render(): string {
    const cell = this.controller.cell;
    let result = '';
    const piece = cell.getPiece();
    const background = cell.color === 'black' ? '#D18B47' : '#FFCE9E';
    const icon = piece ? Icons[piece.getType()][piece.getColor()] : ' ';
    result = this.engine.renderBox({
      column: cell.colIndex,
      row: cell.rowIndex,
      background,
      text: icon,
    });

    // if (piece === null) {
    //   if (!cell.isHighlighted()) {
    //     result = this.engine.renderBox({
    //       column: cell.file,
    //       row: cell.rank,
    //       background,
    //       text: ' ',
    //     });
    //   } else {
    //     result = '× ';
    //   }
    // } else {
    //   const icon = Icons[piece.getType()][piece.getColor()];
    //   if (cell.isHighlighted()) {
    //     result = '×' + icon;
    //   } else result = icon + ' ';
    // }
    return result;
  }
}
