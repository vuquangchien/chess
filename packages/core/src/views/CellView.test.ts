import {Cell} from '../models/Cell';
import {CellView} from './CellView';
import {CellViewController} from './CellViewController';
import {Piece} from '../models/Piece';

describe('CellView', () => {
  it('should render empty cell', () => {
    const cell = Cell.fromName('a1');
    const cellView = new CellView(new CellViewController(cell));
    expect(cellView.render()).toMatchSnapshot();
  });
  it('should render cell with piece', () => {
    const cell = Cell.fromName('a1');
    cell.setPiece(Piece.fromName('b'));
    const cellView = new CellView(new CellViewController(cell));
    expect(cellView.render()).toMatchSnapshot();
  });
  // render white cell
  it('should render white cell', () => {
    const cell = Cell.fromName('a8');
    const cellView = new CellView(new CellViewController(cell));
    expect(cellView.render()).toMatchSnapshot();
  });
  // render highlighted cell
  it('should render highlighted cell', () => {
    const cell = Cell.fromName('a8');
    cell.highlight();
    const cellView = new CellView(new CellViewController(cell));
    expect(cellView.render()).toMatchSnapshot();
  });
  // render highlighted cell with piece
  it('should render highlighted cell with piece', () => {
    const cell = Cell.fromName('a8');
    cell.highlight();
    cell.setPiece(Piece.fromName('b'));
    const cellView = new CellView(new CellViewController(cell));
    expect(cellView.render()).toMatchSnapshot();
  });
});
