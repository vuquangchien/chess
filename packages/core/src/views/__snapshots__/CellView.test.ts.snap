// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellView should render cell with piece 1`] = `"♝ "`;

exports[`CellView should render empty cell 1`] = `"■ "`;

exports[`CellView should render highlighted cell 1`] = `"× "`;

exports[`CellView should render highlighted cell with piece 1`] = `"×♝"`;

exports[`CellView should render white cell 1`] = `"□ "`;
